src/query_builder.ts

```diff
@@ -1,26 +1,88 @@
 import { API } from './api'
 
-type Filter =
-  | string
-  | string[]
-  | Record<string, number | number[] | string | string[]>
+type BaseValueFilter = string | string[] | number | number[]
+type BaseRangeFilter = {
+  gte?: number
+  lte?: number
+  gt?: number
+  lt?: number
+}
+type FilterFieldValue = BaseValueFilter | BaseRangeFilter
+type FilterMatchField = { match: { [field: string]: BaseValueFilter } }
+type FilterTermsField = { terms: { [field: string]: BaseValueFilter } }
+type FilterRangeField = { range: { [field: string]: BaseRangeFilter } }
+
+type FilterField = FilterMatchField | FilterRangeField | FilterTermsField
+
+enum FacetType {
+  terms = 'terms',
+  stats = 'stats',
+}
+
+interface FacetParam {
+  field: string
+  size?: number
+  type: FacetType
+  disjunctive?: boolean
+}
+
+const mapFacetTypeToElastic: { [key: string]: 'terms' | 'range' } = {
+  [FacetType.terms]: 'terms',
+  [FacetType.stats]: 'range',
+}
+
+interface Params {
+  params: {
+    _es_filters?: { bool: { must: FilterField[] } }
+    query?: string
+    result_fields?: string[]
+    search_fields?: string[]
+    size?: number
+    from?: number
+  }
+}
 
 export class QueryBuilder {
-  private params = { params: { query: '' } }
+  private readonly facets: Record<string, FacetParam>
+  params: Params = { params: {} }
```

Comment:
nitpick: does params need to be nested?

Reply:
Not, I did it for tests, and private field issue, and forgot to return it back